new Dirac kernels that reads in multiple point sources a vpp

lynnmunday opened the pull request:
The tests for the ConstantPointSource were changed to have multiple point source blocks and the the new tests for the vectorPostprocessorPointSource exodiff against these gold files.  

The vectorPostprocessorPointSource reads in values and points from a vpp.  The values for the points can be updated as the vpp changes in a transient simulation.  This will be useful for transferring vpp data between multiapps. 

closes #15831

lynnmunday commented:
@dschwen this is the corrected PR

loganharbour commented:
Please push a change to this to forcefully re-trigger testing (full testing is now available).

You can do the following:
```
git commit --amend
```
and save the amend message that comes up without changing it (by typing `:wq` if in vim). This will forcefully change the commit hash of the last commit, which will trigger a change. Then forcefully push:
```
git push --force
```

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/582471/) on 72bedae wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15835/site)

This comment will be updated on new commits.

lynnmunday commented:
@dschwen will you review?

dschwen commented on framework/include/dirackernels/ConstantPointSource.h:
```diff
@@ -19,22 +20,30 @@ template <>
 InputParameters validParams<ConstantPointSource>();
 
 /**
- * TOOD
```
🎺

lynnmunday replied:
I can't find this in the latest version.  Maybe I fixed it.

lynnmunday replied:
spelling?

dschwen replied:
trumpet noise

dschwen commented on framework/include/dirackernels/ConstantPointSource.h:
```diff
@@ -19,22 +20,30 @@ template <>
-
+  /// Need initialSetup because vectorpostprocessor will not be available at construction
+  virtual void initialSetup() override;
+  /// map to associate points with their value in computeQpResidual
```
these comments are probably better placed in the .C file as they do not appropriately document these member methods

lynnmunday replied:
I reverted the vpp part of this so now everything is in the constructor and these comments were removed

dschwen commented on framework/src/dirackernels/ConstantPointSource.C:
```diff
@@ -17,39 +19,167 @@ InputParameters
+      "vector_postprocessor",
+      "The name of the VectorPostprocessor containing positions and corresponding values with "
+      "columns named x y z value");
+
```
I think this crams 3 objects worth of functionality into one object. I'd rather see this refactored into `FilePointSource` and `VectorPostprocessorPointSource`

dschwen replied:
I'd begrudgingly accept the file reader option in Constant point source (there is precedent for this). But I don't really like it.

The VPP is a hard no for me, as that is not "constant" at all anymore.

lynnmunday replied:
I reverted the vpp part of this commit.  It should be its own type of point source.  I had implemented it wrong and it would have been constant.  I need to make the point source change as the vpp changes which the previous commit did not do.
The file reader is still part of constantPointSource

lynnmunday replied:
I made a new object called FileConstantPointSource and moved the file reading to that.  I reverted all of the changes I made to ConstantPointSource so constantPointSource will only read in a single point source per block.  I need to move the vpp part to another object but that isn't constant so I had implemented it wrong.

dschwen commented on framework/src/dirackernels/ConstantPointSource.C:
```diff
@@ -17,39 +19,167 @@ InputParameters
+      if (dim > 2)
+      {
+        pt(2) = point_param[dim * i + 2];
+      }
```
too many curly braces

dschwen commented on framework/src/dirackernels/ConstantPointSource.C:
```diff
@@ -17,39 +19,167 @@ InputParameters
+  if (values.size() == 0 || point_param.size() == 0)
+    mooseError("ConstantPointSource: No values or points created by input or position_value_file.");
+
+  size_t dim = point_param.size() / values.size();
```
check for remainder

dschwen commented on framework/src/dirackernels/ConstantPointSource.C:
```diff
@@ -17,39 +19,167 @@ InputParameters
+  for (auto & point : _point_to_value)
+  {
+    addPoint(point.first);
+    ++index;
```
what do you use `index` for?

lynnmunday replied:
I don't know?

lynnmunday committed:
vectorPostprocessorPointSource that will read in point source from a vpp.

lynnmunday committed:
2d and 3d constant vpp point sources that match constantPointSource results

lynnmunday committed:
renaming files for consistency.  Added multiple ConstantPointSources to each file for comparing results to fileConstantPointSource results.

ref #15831

lynnmunday committed:
reformatted input file paramter name

lynnmunday committed:
added transient test where the point source changes as the vpp changes.  Also added better error message.

lynnmunday committed:
add documentation for VectorPostprocessorPointSource

Also added
-better error checking for when nothing is read in.
-added vpp vector names as member data.
-Added csv file for csvFilereader test.
-fixed names of gold files to be the same as the input file

moosebuild commented:
Job [App tests](https://civet.inl.gov/job/592680/) on f5766bb : invalidated by @lindsayad

moosebuild commented:
Job [App tests (Non Conda)](https://civet.inl.gov/job/592693/) on f5766bb : invalidated by @lindsayad

moosebuild commented:
Job [Non unity build](https://civet.inl.gov/job/592695/) on f5766bb : invalidated by @lindsayad

lynnmunday commented:
@dschwen I reverted the old code and started over with separate classes. Would you review it again?

lynnmunday commented:
@dschwen will you review this

dschwen commented on framework/include/dirackernels/VectorPostprocessorPointSource.h:
```diff
@@ -0,0 +1,41 @@
+
+/**
+ * A VectorPostprocessorPointSource DiracKernel is used to create constant point sources.
+ * Coordinates and values for the point source can be given in the input or read from file
```
update this

lynnmunday replied:
fixed

dschwen commented on framework/include/dirackernels/FileConstantPointSource.h:
```diff
@@ -0,0 +1,43 @@
+/**
+ * A FileConstantPointSource DiracKernel is used to create constant point sources.
+ * Coordinates and values for the point source can be given in the input or read from file
+ */
```
only from file

lynnmunday replied:
I deleted this class since the same thing can be done with a combination of a vpp Dirackernel that reads in a csv vpp.  Their is a test for this.

dschwen commented on test/tests/dirackernels/vectorPostprocessor_point_source/3d_vpp.i:
```diff
@@ -0,0 +1,67 @@
+[]
+
+[Outputs]
+  file_base = 3d_out
```
let's remove all those file_base params, they make it harder to see which output corresponds to which input

lynnmunday replied:
This is so the output file can be diffed with the gold file from constant point source.

dschwen replied:
my preference would be a symlink in the gold directory, but I think we may not have an "official" stance on that

lynnmunday replied:
I changed these to not use the file_base and to use a symlink with a different name than the file being linked.

dschwen commented on framework/src/dirackernels/VectorPostprocessorPointSource.C:
```diff
@@ -0,0 +1,80 @@
+void
+VectorPostprocessorPointSource::addPoints()
+{
+  const auto xcoord(getVectorPostprocessorValue(
```
this could be gotten as a member in the initializer list, or is there a reason fro re-getting this each time step?

dschwen replied:
If that was the case we couldn't have moving point sources

lynnmunday replied:
I made the names member data and they are added to the initializer list.

lynnmunday commented:
@dschwen are all of the issues resolved?

dschwen commented on framework/src/dirackernels/VectorPostprocessorPointSource.C:
```diff
@@ -0,0 +1,83 @@
+
+registerMooseObject("MooseApp", VectorPostprocessorPointSource);
+
+defineLegacyParams(VectorPostprocessorPointSource);
```
Not needed for new objects anymore, just FYI

lynnmunday replied:
I tried not using it and I get: 
dirackernels/vectorPostprocessor_point_source.2dTransient: dyld: Symbol not found: __Z11validParamsI30VectorPostprocessorPointSourceE15InputParametersv
dirackernels/vectorPostprocessor_point_source.2dTransient:   Referenced from: /Users/mundlb/projects/moose/framework/libmoose-opt.0.dylib
dirackernels/vectorPostprocessor_point_source.2dTransient:   Expected in: flat namespace

lynnmunday commented:
@dschwen I made the changes you recommended, would you look it over again.  I can't get rid of the definelegacyparam
